pkg/parser/parser.go

```diff
@@ -0,0 +1,291 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package parser
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"strings"
+
+	"github.com/spf13/afero"
+	corev1 "k8s.io/api/core/v1"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes"
+
+	apiextensionsv1alpha1 "github.com/crossplane/crossplane/apis/apiextensions/v1alpha1"
+	pkgmetav1alpha1 "github.com/crossplane/crossplane/apis/pkg/meta/v1alpha1"
+)
+
+// Parser is a package parser.
+type Parser interface {
+	Parse(context.Context, ...BackendOption) (*Package, error)
+}
+
+// parser is a Parser implementation with pluggable backends.
+type parser struct {
+	backend Backend
+}
+
+// New returns a new parser.
+func New(backend Backend) Parser {
+	return &parser{
+		backend: backend,
+	}
+}
+
+// BackendOption modifies the parser backend. Backends may accept options at
+// creation time, but must accept them at initialization.
+type BackendOption func(Backend)
+
+// Backend provides a source for the parser.
+type Backend interface {
+	Init(context.Context, ...BackendOption) (io.ReadCloser, error)
+}
+
+// Package includes fields that all meaningful parsers should support.
+type Package struct {
+	provider                     *pkgmetav1alpha1.Provider
+	configuration                *pkgmetav1alpha1.Configuration
+	customResourceDefinitions    map[string]*apiextensions.CustomResourceDefinition
+	compositeResourceDefinitions map[string]*apiextensionsv1alpha1.CompositeResourceDefinition
+	compositions                 map[string]*apiextensionsv1alpha1.Composition
+}
```

Comment:
I suspect this same concern could arise for the `*Provider` and `*Configuration` types too. I almost wonder whether we should only have two methods:

```go
type PackageMeta interface {
    // Some methods that all Configurations and Package types will have?
}

func (p *Package) GetMeta() PackageMeta {
  // Returns either a Provider or a Config
}

func (p *Package) GetObjects() []runtime.Object {}
```

Reply:
I was initially thinking this would be difficult to work with in the package revision controller, but considering also that we could validate the package beforehand (i.e. for a `Provider`, make sure that the only objects are CRDs and and a `Provider` meta type), then we could pass the validator into `NewReconciler`, parse the package, run the validator, then just do `p.GetObjects()` and know that we are dealing with a valid `Provider` package. Let me rework this and see how it feels :+1:

Reply:
Thanks! I think the advantage of moving to `runtime.Object` is that it decouples this package from knowledge of any particular API types, and instead makes it responsible for:

1. Reading YAML documents (from a stream or dir).
1. Converting those YAML documents into a type that distinguishes the package metadata document (Package or Configuration) from the payload (XRDs, CRDs, etc).

This way the package revision controller could be updated to extract the types it needs from `GetObjects()` via a filter function. Something like:

```go
GetXRDs(in []runtime.Object) []v1alpha1.CompositeResourceDefinition {
    // Use type assertion to return only the Objects that are XRDs.
} 
```

Reply:
I think there's more value than a generic YAML parser in this package - it also contains the logic to read the YAML from various Crossplane specific sources (package dirs, pod logs, etc), loads Kubernetes objects specifically, and can categorise them into package metadata or payload. I think this is valuable, probably more-so than being able to extract and present specific object types.